Added Widevine Base License Acquisition + FairPlay support

rnrneverdies opened the pull request:
We also included a new sample for FairPlay.

azurecla commented:
Hi __@rnrneverdies__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

azuresdkci commented:
Can one of the admins verify this patch?

yaqiyang commented:
Thanks for the contribution. I will take a look in the next few days.

phansys commented on .gitignore:
```diff
@@ -9,3 +9,4 @@ vendor
 nbproject
 phpcs-checkstyle.tmp
 composer.phar
+.vscode/
```
Why is this change required? Environment specific files/directories should be added to your local ignore list: https://help.github.com/articles/ignoring-files/#create-a-local-gitignore

phansys commented on examples/MediaServices/vodworkflow_aes.php:
```diff
@@ -331,7 +331,7 @@ function publishEncodedAsset($restProxy, $encodedAsset)
 
     // 6.4 Create a Smooth Streaming base URL
-    $stremingUrl = $locator->getPath().$manifestFile->getName().'/manifest';
+    $stremingUrl = $locator->getPath().$manifestFile->getName()."/manifest";
```
How is this change related to Widevine / FairPlay?

rnrneverdies replied:
It's not, but we went forced to applying many patches because someone has refactored the entire repository, this one seems that is stowaway from the older branch.

rnrneverdies replied:
fixed

phansys commented on src/MediaServices/Models/AssetDeliveryPolicyConfigurationKey.php:
```diff
@@ -96,6 +96,35 @@ class AssetDeliveryPolicyConfigurationKey
      */
     const WIDEVINE_LICENSE_ACQUISITION_URL = 7;
 
+	/**
```
Please, fix indentation.

rnrneverdies replied:
fixed

phansys commented on src/MediaServices/Templates/FairPlayConfiguration.php:
```diff
@@ -0,0 +1,76 @@
+        $template = new FairPlayConfiguration();
+        $template->ASkId = $askId;
+        $template->ContentEncryptionIV = $contentIv;
+		$template->FairPlayPfx = $certString;
```
Please, fix indentation.

rnrneverdies replied:
fixed

phansys commented on src/MediaServices/Templates/FairPlayConfiguration.php:
```diff
@@ -0,0 +1,76 @@
+    }
+   
+}
+
```
Please, remove surplus line breaks at EOF.

rnrneverdies replied:
fixed

yaqiyang commented:
I got two problems when I ran your code against the unit tests.
First, I am getting this error.
  <error type="PHPUnit_Framework_Error_Warning">Tests\unit\WindowsAzure\MediaServices\MediaServicesRestProxyTest::testCreateContentKey
mcrypt_encrypt(): Size of key is too large for this algorithm

D:\Repos\azure-sdk-for-php\src\MediaServices\Models\ContentKey.php:427
D:\Repos\azure-sdk-for-php\src\MediaServices\Models\ContentKey.php:453
D:\Repos\azure-sdk-for-php\tests\unit\WindowsAzure\MediaServices\MediaServicesRestProxyTest.php:1294
</error>
It means mcrypt_encrypt() is failing with your change at src\MediaServices\Models\ContentKey.php:427. Does this change require configuration changes?

Second, you changed src/MediaServices/Templates/TokenRestrictionTemplateSerializer.php which also cause a test to fail (Tests\unit\WindowsAzure\MediaServices\Templates\TokenRestrictionTemplateSerializerTest::testSerializerSWT) because the serialized string is now different. Will this change cause any regressions for existing users?

Thanks
Yaqi Yang

rnrneverdies commented:
Hi @yaqiyang, sorry for the delay.

I'm going to take a look on your comments asap and will let you know. 

Thanks.

rnrneverdies commented:
Hi @yaqiyang, I've updated this branch with the following changes:

 - Reverted the changes on the token restriction template to keep it backwards compatible.
 - Updated content key checksum generation algorithm to make it optional and to keep it backwards compatible.

Please let me know if you have additional feedback.

Thanks.

yaqiyang commented:
Thanks. I will take a look once I come back from vacation next week.
